### PR TITLE
Add post: bert-model

### PR DESCRIPTION
**Pull request by abhishalya**

### CODE REVIEW

**Comment by abhishalya** [comment]
I've added netlify to this repo to get deploy previews for easy review.

**Review by logankilpatrick — APPROVED** [review]
Wow, great blog post. Very well written. Keep up the great work.

**Comment by abhishalya** [comment]
Thank you @logankilpatrick :smile:

File: _posts/2019-12-10-bert-model.md [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+
+In BERT, we don't represent each word as a vector (as we've been doing since
+years in NLP) instead we represent each 'subword token' as a vector. That is,
+whenever the model finds an unknown word it tries to split the word into
+sub-pieces like the word root, prefixes, etc  and use all the representational
+vectors of each sub-pieces together to represent the unfamiliar word.
+
+![bert-work-1](../assets/images/bert-work-1.png)
+
+Finally, we take the output of the last encder block as our final representation.
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md]: encoder

Comment (aviks) [inline _posts/2019-12-10-bert-model.md:55]: encoder

File: _posts/2019-12-10-bert-model.md [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+We can just feed the representation of our old NLP model to BERT and get a much
+better performance. This means that the BERT isn't as complicated as some of the
+other NLP models. However, the BERT model is VERY large. As an instance,
+for the Aamzon Review dataset I wasn't able to train 10 reviews on my 6 GB GPU.
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:59]: Do you mean 10 reviews? or 10 batches?

File: _posts/2019-12-10-bert-model.md [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+Google almost every new term I had discovered while reading. So, I spent time
+reading and understanding and after I had a basic idea, I was ready to look at
+the example code.
+
+The Transformers.jl repository already has few examples on BERT. Here we are
+working on Sentiment Analysis and hence we need to construct a classifier. We
+will now see the code and the output for the model I constructed. Its a bit
+long, and I won't be able to explain each and every section of the code. But,
+I'll cover up the important ones.
+
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:93]: Say a little about what you are classifiying. What are you trying to predict? What are the lables?

File: _posts/2019-12-10-bert-model.md [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+```
+Of course, you can change these as per your requirement. I've set Batch to 2
+to save some memory and Epoch to 1 to save some time. Epoch being one just
+would mean that there will be a single iteration of weights update.
+
+Now, lets get the pretrained model. Here the downloading extracting the model
+will be taken care of by Transformers. All you need to do is following.
+
+```jl
+const _bert_model, wordpiece, tokenizer = pretrain"Bert-uncased_L-12_H-768_A-12"
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:169]: explicitly mention here that loading the pretrained model gives you three things -- the model weights, the vocabulary, and the tokenizer. It's easy for the reader to miss this and be confused later

File: _posts/2019-12-10-bert-model.md [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+
+To prepare it the way we require, we would need to add few more functions. This
+will create an iterator like object for our data.
+
+```jl
+using DelimitedFiles
+
+function gen_train()
+    sets = readdlm("new_pre_train.tsv", '\t', String, quotes = false)
+    return [selectdim(sets, 2, i) for i = (1,2)]
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:223]: Is it possible to explain what this line does? what is the output you are looking for (beyond the fact that it's and iterator? Iterator over what? ie, what element does iterating over the thing return?

File: _posts/2019-12-10-bert-model.md [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+end
+
+datas = gen_train() # This will be used in the `train!()` function
+```
+
+```jl
+const vocab = Vocabulary(wordpiece)
+
+markline(sent) = ["[CLS]"; sent; "[SEP]"]
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:232]: Is it possible to explain what this does?

File: _posts/2019-12-10-bert-model.md [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+    segment = fill!(similar(tok), 1)
+    
+    label = onehotbatch(batch[2], labels)
+    return (tok = tok, segment = segment), label, mask
+end
+
+preprocess(get_batch(datas, 4))
+```
+
+Here basically what we do it get mask for a sentence, on a way that we don't
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:247]: 'on a way'? maybe reword?

File: _posts/2019-12-10-bert-model.md [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+        res[i] = x
+        i >= n && break
+    end
+    isassigned(res, n) ? batched(res) : nothing
+end
+```
+
+### Step 3: Training and testing the model
+
+First, lets define the `bert_model` and the `clf`.
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:266]: explicitly mention that you are running it on the GPU

File: _posts/2019-12-10-bert-model.md [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+            update!(opt, ps, grad)
+            i%16==0 && @show al/i
+        end
+
+        test()
+    end
+end
+```
+
+After looping through all of the data, we will move towards testing it:
```
Comment (aviks) [inline _posts/2019-12-10-bert-model.md:333]: This part will make more sense if you've explained above what label you are classifying. Once you have that, maybe refer back to that.

**Review by aviks — COMMENTED** [review]
I've left some comments. You don't have to agree to everything, but hopefully you'll consider them.

File: _posts/2019-12-10-bert-model.md (hunk 2 above)
Comment (abhishalya) [reply _posts/2019-12-10-bert-model.md:59]: 10 reviews

File: _posts/2019-12-10-bert-model.md (hunk 3 above)
Comment (abhishalya) [reply _posts/2019-12-10-bert-model.md:93]: Sure, I'll modify this

File: _posts/2019-12-10-bert-model.md (hunk 7 above)
Comment (abhishalya) [reply _posts/2019-12-10-bert-model.md:247]: Yep :sweat_smile:

**Comment by abhishalya** [comment]
Thank you @aviks for reviewing the post. Since this PR is already merged, I'll create another one to incorporate all of the changes requested. I'll do this once my current task is done, and will let you know on slack.
